<a href="https://colab.research.google.com/github/mariacgomez-tech/Analitica-de-Negocios/blob/main/HeartDisiese.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Caso de estudio**
Se quiere implementar un modelo de árbol de decisión para pronosticar si las personas que hacen parte de esta base de datos van a desarrollar una enfermedad cardiaca. Para este proceso vamos a utilizar las variables:

**Age**: Indica el número de años que posee la persona al momento del estudio.

**Systolic**: Indica la presión máxima en las arterias cuando el corazón se contrae. (Indica la fuerza con la que el corazón bombea sangre al cuerpo). Este factor es más determinante al momento de identificar un posible accidente cardiovascular.


**Diastolic**: Indica la presión mínima en las arterias cuando el corazón se relaja entre latidos.


**BMI**: Es el índice de masa corporal de la persona al momento del estudio. Indica si la persona tiene un peso saludable en relación con su altura.

0. Se procede con la carga de librerías de trabajo

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. Se procede con la carga de datos de trabajo.

In [ ]:
nxl= '/content/sample_data/2. HeartDisease.xlsx'
XDB1= pd.read_excel(nxl, sheet_name= 0)
XDB2= pd.read_excel(nxl, sheet_name= 1)

XD1=XDB1.iloc[:,[1,2,3,4]]
yd1=XDB1.iloc[:,0]
display(XD1)
display(yd1)



,Age,Systolic,Diastolic,BMI
0,44,112,111,17
1,55,128,90,27
2,47,131,94,26
3,31,151,104,17
4,65,148,117,17
...,...,...,...,...
175,43,119,76,25
176,63,107,113,40
177,47,116,61,27
178,70,110,88,24


,Disease
0,0
1,1
2,1
3,0
4,1
...,...
175,0
176,1
177,0
178,0


2. Se procede con la implementación del modelo de árbol

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4) #Entrenamos el árbol con base histórica de datos
mar.fit(XD1,yd1)

ydp1=mar.predict(XD1) #Predicción del modelo
#Se construye la matriz de confusión
cm=confusion_matrix(yd1,ydp1)
display(cm)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

#Métricas de desempeño
Ex=(VP+VN)/len(XD1) #1. Exactitud: Comportamiento general
print('La Exactitud es: ',Ex)
Sen=VP/(VP+FN) #2. Sensibilidad: Como se comporta pronosticando PreAprobados
print('La Sensibilidad es: ',Sen)
Spe=VN/(VN+FP) #3. Especificidad: Como se comporta frente a los negativos
print('La Especificidad es: ',Spe)
Pre=VP/(VP+FP) #4. Precisión: Como se comporta pronosticando PreAprobados
print('La Precisión es: ',Pre)
PreNeg=VN/(VN+FN) #5. Precisión Negativa: Como se comporta pronosticando Negativos
print('La Precisión Negativa es: ',PreNeg)

array([[63, 10],
       [20, 87]])

La Exactitud es:  0.8333333333333334
La Sensibilidad es:  0.8130841121495327
La Especificidad es:  0.863013698630137
La Precisión es:  0.8969072164948454
La Precisión Negativa es:  0.7590361445783133


**Análisis de resultados**

De la base de datos histórica se observa un total de 180 registros, de los cuales 73 corresponden a personas sin enfermedad cardiaca  y 107 corresponden a personas con enfermedad cardiaca.

De los 73 casos sin enfermedad, el modelo pronosticó correctamente un total de 63 personas, mientras que de los 107 casos con enfermedad, el modelo clasificó correctamente 87 registros. El modelo logra identificar en un 86.30% a las personas sin enfermedad (63/73), mientras que logra identificar en un 81.31% a las personas con enfermedad (87/107).

Con respecto a las métricas, el modelo alcanzó una exactitud del 83.33%, lo que indica un muy buen comportamiento general del modelo frente a la clasificación en las categorías. Se destacan especialmente la especificidad (86.30%) y la precisión (89.69%), las cuales se ubican ampliamente por encima del umbral del 75% definido como referencia para este tipo de modelos de clasificación. La sensibilidad (81.31%) también presenta un desempeño sólido, mostrando una adecuada capacidad para detectar pacientes con enfermedad cardiaca. La precisión negativa (75.90%) se mantiene dentro de un rango aceptable, evidenciando un modelo equilibrado entre ambas clases.

De acuerdo con el árbol de decisión se identifican varios nodos puros (gini = 0). Al analizar el lado derecho del árbol, se destaca especialmente
un nodo puro (2/0) bajo la regla de decisión Age > 69.5, en el cual todos los individuos son clasificados como personas sin enfermedad. Así mismo, se destaca el nodo puro (0/35) bajo la regla Age > 54.5 AND Age ≤ 69.5 AND Diastolic > 85.5, donde el 100% de los individuos presentan enfermedad cardiaca, evidenciando un perfil de alto riesgo claramente diferenciado.
Es importante destacr la existencia de más nodos puros en el lado izquierdo del árbol.





3. Despliegue del árbol de decisión

In [ ]:
from sklearn.tree import export_graphviz #Exporta los datos a un gráfico
from pydotplus import graph_from_dot_data #Es un graficador

vs=["Age", "Systolic", "Diastolic", "BMI"] #Títulos del árbol
dot_data=export_graphviz(mar, feature_names=vs) #Exportar de números a gráfico en pdf
graph=graph_from_dot_data(dot_data) #Graficamo el árbol
graph.write_png('arbol1.png') #Guardamos el árbol en pdf

True

4. Pronósticar con 'Datos Pronóstico'

In [ ]:
XD2=XDB2.iloc[:,[1,2,3,4]]
ydp2=mar.predict(XD2) #Predicción del modelo
display(ydp2)

array([0, 0, 0, 1, 1, 1, 1, 1])

**Análisis Pronóstico**

De los 8 individuos evaluados, el modelo pronostica que 3 personas (3/8 = 37.5%) no desarrollarán enfermedad cardiaca, mientras que 5 personas (5/8 = 62.5%) presentan riesgo de desarrollarla.

El pronóstico evidencia que la mayoría de las personas analizadas corresponden a perfiles que, según las reglas aprendidas por el modelo, presentan características asociadas a mayor probabilidad de enfermedad.